File: Week 6/solutions/F/F_keyboards.java
import java.util.*;
import java.io.*;

class Main{
    static final int INF = (1 << 30) - 1;
    static final int N = (int)1e3 + 5;

    static int[][] dp = new int[N][N];
    static int n, m;
    static HashSet<Integer>[] notes = new HashSet[N];
    static ArrayList<Integer>[] played_by = new ArrayList[N];
    static int[] a = new int[N];
    static int solve(int i, int j){
        // i = idx of current note, j = current instrument
        // if j == 0, we need to switch instruments
        // we can store which instruments play which notes and iterate, and just recursively call again
        if(i == m + 1) return 0;
        int ans = dp[i][j];
        if(ans != -1) return ans;
        if(j == 0){
            ans = INF;
            for(int k : played_by[a[i]]){
                ans = Math.min(ans, solve(i + 1, k));
            }
            ans += 1; // +1 for switching instruments
        }else if(notes[j].contains(a[i])){
            // current instrument can play current note; don't switch
            ans = solve(i + 1, j);
        }else{
            // current instrument doesn't play this note, so we need to switch
            ans = solve(i, 0);
        }
        return dp[i][j] = ans;
    }
    public static void main(String[] args) throws IOException{
        BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(f.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        for(int i = 1; i <= n; i++){
            notes[i] = new HashSet<>();
            st = new StringTokenizer(f.readLine());
            int k = Integer.parseInt(st.nextToken());
            for(; k > 0; k--){
                int x = Integer.parseInt(st.nextToken());
                notes[i].add(x);
                if(played_by[x] == null) played_by[x] = new ArrayList<>();
                played_by[x].add(i);
            }
        }
        st = new StringTokenizer(f.readLine());
        for(int i = 1; i <= m; i++){
            a[i] = Integer.parseInt(st.nextToken());
        }
        for(int i = 0; i < N; i++){
            Arrays.fill(dp[i], -1);
        }
        System.out.println(solve(1, 0) - 1);
    }
}

File: Week 6/solutions/F/F_keyboards.py

INF = (1 << 30) - 1
N = 1000 + 5

dp = [[-1 for _ in range(N)] for _ in range(N)]
notes = [set() for _ in range(N)]
played_by = [[] for _ in range(N)]
a = []
def solve(i,j):
    # i = idx of current note, j = current instrument
    # if j == 0, we need to switch instruments
    # we can store which instruments play which notes and iterate, and just recursively call again
    if i == m + 1: return 0
    if dp[i][j] != -1: return dp[i][j]
    if j == 0:
        dp[i][j] = INF
        for k in played_by[a[i]]:
            dp[i][j] = min(dp[i][j], solve(i + 1, k))
        dp[i][j] += 1 # +1 for switching instruments
    elif a[i] in notes[j]:
        # current instrument can play current note; don't switch
        dp[i][j] = solve(i + 1, j)
    else:
        # current instrument doesn't play this note, so we need to switch
        dp[i][j] = solve(i, 0)

    return dp[i][j]

n, m = map(int, input().split())
for i in range(1, n + 1):
    ls = list(map(int, input().split()))
    k = ls[0]
    for j in range(1, k + 1):
        notes[i].add(ls[j])
        played_by[ls[j]].append(i)

a = [0] + list(map(int, input().split()))
print(solve(1, 0) - 1)